File: 2025/L/answers/java/Solution.java
import java.util.*;
import java.io.*;

public class Solution {
    static class Interval implements Comparable<Interval> {
        double l, r;
        public Interval(double l, double r) { this.l = l; this.r = r; }
        @Override
        public int compareTo(Interval other) {
            if (Math.abs(this.l - other.l) > 1e-11) return Double.compare(this.l, other.l);
            return Double.compare(this.r, other.r);
        }
    }

    static int[] bit;
    static int mSize;

    static void update(int i, int val) {
        for (; i <= mSize; i += i & -i) bit[i] += val;
    }

    static int query(int i) {
        int res = 0;
        for (; i > 0; i -= i & -i) res += bit[i];
        return res;
    }

    static int findFirstGE(double[] arr, double target) {
        int low = 0, high = arr.length - 1;
        int res = arr.length;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (arr[mid] >= target - 1e-13) {
                res = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return res;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.`in`);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        double a = sc.nextDouble();
        double b = sc.nextDouble();

        List<Interval> intervals = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            double m = sc.nextDouble();
            double c = sc.nextDouble();
            double disc = m * m - 4.0 * a * (b - c);
            if (disc > 1e-11) {
                double sd = Math.sqrt(disc);
                double l = (m - sd) / (2.0 * a);
                double r = (m + sd) / (2.0 * a);
                intervals.add(new Interval(l, r));
            }
        }

        if (intervals.isEmpty()) {
            System.out.println(0);
            return;
        }

        Collections.sort(intervals);

        double[] allR = new double[intervals.size()];
        for (int i = 0; i < intervals.size(); i++) allR[i] = intervals.get(i).r;
        Arrays.sort(allR);
        
        int m = 0;
        if (allR.length > 0) {
            m = 1;
            for (int i = 1; i < allR.length; i++) {
                if (Math.abs(allR[i] - allR[m-1]) > 1e-11) {
                    allR[m++] = allR[i];
                }
            }
        }
        
        double[] finalAllR = Arrays.copyOf(allR, m);
        mSize = m;
        bit = new int[mSize + 1];

        long ans = 0;
        double eps = 1e-11;
        for (Interval in : intervals) {
            int idxL = findFirstGE(finalAllR, in.l + eps);
            int idxR = findFirstGE(finalAllR, in.r - eps); 
            // idxL is first index where R_j > L_i + eps
            // idxR is first index where R_j >= R_i - eps
            // We want indices j in [idxL, idxR-1]
            
            if (idxR > idxL) {
                ans += query(idxR) - query(idxL);
            }
            
            int rIdx = findFirstGE(finalAllR, in.r - 1e-13);
            update(rIdx + 1, 1);
        }
        System.out.println(ans);
    }
}

File: 2025/L/answers/kt/Solution.kt
import java.util.*

class Interval(val l: Double, val r: Double) : Comparable<Interval> {
    override fun compareTo(other: Interval): Int {
        if (Math.abs(l - other.l) > 1e-11) return l.compareTo(other.l)
        return r.compareTo(other.r)
    }
}

var bit = IntArray(0)
var mSize = 0

fun update(index: Int, value: Int) {
    var i = index
    while (i <= mSize) {
        bit[i] += value
        i += i and -i
    }
}

fun query(index: Int): Int {
    var i = index
    var res = 0
    while (i > 0) {
        res += bit[i]
        i -= i and -i
    }
    return res
}

fun findFirstGE(arr: DoubleArray, target: Double): Int {
    var low = 0
    var high = arr.size - 1
    var res = arr.size
    while (low <= high) {
        val mid = (low + high) / 2
        if (arr[mid] >= target - 1e-13) {
            res = mid
            high = mid - 1
        } else {
            low = mid + 1
        }
    }
    return res
}

fun main() {
    val sc = Scanner(System.`in`)
    if (!sc.hasNextInt()) return
    val n = sc.nextInt()
    val a = sc.nextDouble()
    val b = sc.nextDouble()

    val intervals = mutableListOf<Interval>()
    for (i in 0 until n) {
        val m = sc.nextDouble()
        val c = sc.nextDouble()
        val disc = m * m - 4.0 * a * (b - c)
        if (disc > 1e-11) {
            val sd = Math.sqrt(disc)
            val l = (m - sd) / (2.0 * a)
            val r = (m + sd) / (2.0 * a)
            intervals.add(Interval(l, r))
        }
    }

    if (intervals.isEmpty()) {
        println(0)
        return
    }

    Collections.sort(intervals)

    val allR = DoubleArray(intervals.size)
    for (i in intervals.indices) allR[i] = intervals[i].r
    Arrays.sort(allR)

    var m = 0
    if (allR.isNotEmpty()) {
        m = 1
        for (i in 1 until allR.size) {
            if (Math.abs(allR[i] - allR[m - 1]) > 1e-11) {
                allR[m++] = allR[i]
            }
        }
    }

    val finalAllR = allR.copyOf(m)
    mSize = m
    bit = IntArray(mSize + 1)

    var ans = 0L
    val eps = 1e-11
    for (inVal in intervals) {
        val idxL = findFirstGE(finalAllR, inVal.l + eps)
        val idxR = findFirstGE(finalAllR, inVal.r - eps)

        if (idxR > idxL) {
            ans += query(idxR) - query(idxL)
        }

        val rIdx = findFirstGE(finalAllR, inVal.r - 1e-13)
        update(rIdx + 1, 1)
    }
    println(ans)
}
